### PR TITLE
Improve ChecksummedBytes::extend and clarify data integrity guarantee

### PR DESCRIPTION
**Pull request by passaro**
## Description of change

`ChecksummedBytes` maintains a data buffer and a checksum and guarantees that only validated data can be accessed. Transformations such as `split_off`, `extend`, or `slice` (introduced in this change), may trigger a validation (and return an `IntegrityError` on failure), or propagate existing checksum(s) if possible, allowing for later validation.

This change clarifies the data integrity guarantee in the docs and optimizes the `extend` method to avoid re-validation when the checksums for both slices can be combined. It also avoid a redundant buffer allocation.

Also moves `ChecksummedBytes` and related types into a new root module.

## Does this change impact existing behavior?

No.

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and I agree to the terms of the [Developer Certificate of Origin (DCO)](https://developercertificate.org/).

### CODE REVIEW

**Comment by passaro** [comment]
For a clearer diff of the changes in `ChecksummedBytes` see just [commit 711d37d](https://github.com/awslabs/mountpoint-s3/pull/575/commits/711d37ddda701ecd181b3a9586ff0c80dfd6e7f6).

**Review by dannycjones — DISMISSED** [review]
LGTM, one small comment on `split_off` tests.

Let's get a second review on `extend` and `shrink_to_fit`.

File: mountpoint-s3/src/checksums.rs
Comment (jamesbornholt) [inline mountpoint-s3/src/checksums.rs:92]: This is actually a little subtle — it's only correct because `Self::from_bytes(self.curr_slice.clone())` doesn't copy the underlying bytes, and so `self.validate()` is checking the same physical bytes that the new checksum was computed over in `from_bytes`. If there was a copy involved, we'd have to do some extra work to make sure that copy had its checksum computed at the right time.

I think I'd rather not call `from_bytes` here and instead just do the construction manually, so we can be explicit that this transformation is safe because it's not copying. (It's very unlikely but not impossible that some future developer comes along and changes `from_bytes` to copy, without realizing some callers are making a bracketing assumption). We should probably also add a comment on the `curr_bytes` field to capture an invariant that it's always a subslice of `orig_bytes`.

Comment (dannycjones) [reply mountpoint-s3/src/checksums.rs:92]: Thanks for highlighting this, missed this as a rather big assumption.

Agree, I would prefer to do the construction manually.

Comment (dannycjones) [reply mountpoint-s3/src/checksums.rs:92]: >We should probably also add a comment on the curr_bytes field to capture an invariant that it's always a subslice of orig_bytes.

Is there some way we can codify that invariant? For example, not by using `split_off` but instead taking a (real) slice of `orig_bytes` and tying the lifetimes together?

Comment (jamesbornholt) [reply mountpoint-s3/src/checksums.rs:92]: Self-referential lifetimes are pretty annoying (usually require `unsafe`), and `Bytes` is pretty heavily designed around reference counting, so probably not something we can codify that way.

If we wanted an executable check, we could do something janky like:
```rust
let curr_slice = &self.curr_slice[..];
// will panic if curr_slice isn't a subslice of orig_slice
let _ = self.orig_slice.slice_ref(curr_slice);
```
maybe as a debug assert? But possibly not worth the trouble.

Comment (dannycjones) [reply mountpoint-s3/src/checksums.rs:92]: Got it. Plus even if the lifetimes worked, I don't think it would even be `Send` anymore and unusable in async code.

Random idea: maybe we consider storing the bounds/range in place of `curr_slice`, and create the smaller `Bytes` at the last minute when performing `into_bytes`. Looks like an involved refactor though at first glance.

No opinion here, I'm OK keeping it as a documented invariant.

Comment (passaro) [reply mountpoint-s3/src/checksums.rs]: I've added the comment to `curr_slice` for now

**Review by dannycjones — APPROVED** [review]

**Review by jamesbornholt — APPROVED** [review]

**Pull request merged**
